Allow writing modules in TypeScript

dmsnell opened the pull request:
See prior work in #13117, #15390, and #15397

This PR introduces TypeScript into our Babel presets so that we can write in TypeScript and take advantage of the safety and development experience improvements that it offers.

I have updated the _iframe bridge_ for the block editor as a first transformation to not only demonstrate what an update looks like but also because the integrity of that module is relatively important and the introduction of types brings some immediate tangible benefits.

## Review

 - Please review the configuration which enables TypeScript
 - Please share your thoughts on introducing a gradual type system

---

## Is this a big change?

The answer depends. It isn't going to be a big change unless you want to take advantage of it. Existing JavaScript continues to work, though we can run `tsc` to type-check existing code. As @blowery wrote…

> Neat that it can find latent types in the existing JS…[TypeScript] would have saved my bacon a dozen times in the past month alone

Case in point: while updating the iframe bridge I discovered a latent bug that could only exist in rare transient cases. That bug is patched in this PR.

## What is changing?

If you save your file as a `.ts` or a `.tsx` file it will run through TypeScript and you can include types and interfaces. You can import them as well.

Module resolution should stay the same. You won't need to care if the imported module is `.js` or `.ts`. We're using the `@babel/preset-typescript` transform which is altogether different than the method used in #15390. Previously we attempted to have TypeScript compile its code and pass that into `webpack` as a separate loader. In this revision TypeScript is used to remove the type annotations and feed the plain JavaScript code into Babel, essentially the same model that Flow uses. This allows us to compile code with existing errors and we miss out on a couple of features that might be nice to have with TypeScript's codegen, but it simplifies the whole system.

If your editor is setup to support TypeScript you should automatically get autocomplete and inline error warnings for `.ts` and `.tsx` files.

## What benefits does this bring?

 - autocomplete in the IDE
 - inline error warnings
 - greater safety through automatically detecting possible runtime bugs
 - greater communication in the code afforded by expressive semantic types (what kinds of values does this function want? what possible variants of this string category are there? what prop types do I need to send?)
 - more-confident refactoring by highlighting code that breaks when we change an interface or API

## What drawbacks does this bring?

 - possibly slower compile times. I haven't measured this yet, though considering that the entire compile time has dramatically slowed down in the past few months I don't imagine there will be much notice to running code through TypeScript to remove type annotations
 - new syntax to learn. of note we already introduce new syntax via JSX and via a couple Babel transforms. TypeScript is now widely adopted around the web and the additional syntax is minor. TypeScript is a superset of JavaScript, meaning all JavaScript is valid TypeScript too.

It's difficult for me to enumerate many drawbacks. The benefits here are significant for code reliability and developer experience. Just as with our existing JavaScript code, there will be levels to which people take advantage of the type system. For some, more advanced uses will seem confusing for sure. We should consider that type usage is a component of proper code review.

## What's wrong with this PR?

 - [add problem here]

sirreal reviewed (commented):
I would be really happy to see this go in.

I left some notes after looking things over. I don't see any obvious problems. I'm not well versed in TypeScript configuration.

jsnajdr commented on babel.config.js:
```diff
@@ -27,6 +27,7 @@ const config = {
 				exclude: [ 'transform-typeof-symbol' ],
 			},
 		],
+		'@babel/preset-typescript',
```
Nit: the more concise `@babel/typescript` notation is also possible. Babel will infer the `preset-` and `plugin-` prefixes and suffixes.

dmsnell replied:
Cool! I didn't realize that. thanks!

jsnajdr commented on babel.dependencies.config.js:
```diff
@@ -13,6 +13,7 @@ const config = {
 				exclude: [ 'transform-typeof-symbol' ],
 			},
 		],
+		'@babel/preset-typescript',
```
I wouldn't add the preset to the `babel.dependencies.config.js` unless it's actually needed. That will happen once some NPM package gets published with TypeScript sources only.

blowery replied:
This config is only for transpiliing dependencies (ie stuff in node_modules). We shouldn't need to deal with typescript there.

jsnajdr commented on client/gutenberg/editor/calypsoify-iframe.tsx:
```diff
@@ -38,28 +37,57 @@ import { getEditorPostId } from 'state/ui/editor/selectors';
+	siteAdminUrl: string;
+}
+
+type ConnectedProps = ReturnType< typeof mapStateToProps > & typeof mapDispatchToProps;
```
Is this the best state-of-the-art way how to type connected props? I wish it was simpler 🙂

dmsnell replied:
I have no idea. I chose this way without doing any research based on how I felt was most convenient.

At first I was starting to enumerate the connected properties like this:

```ts
interface ConnctedProps {
	siteId: number;
	…
	trashPost: Function;
	…
}
```

Actually, at first I was mashing it all into `Props` but then realized it would be convenient to split them out. Then I felt like this was all pretty redundant because TypeScript is already inferring the types of `mapStateToProps` and `mapDispatchToProps`. Further, because I was hard-coding in those things it seemed fragile, plus I had issues inferring the types of the dispatch functions and I didn't want to widen them.

Using `ReturnType< typeof mapStateToProps >` and `typeof mapDispatchToProps` gave me the information I wanted (a contract for what `Props` are required from the outside as well as a contract for all props) while retaining the flexibility and stronger inference than typing them out myself.

Of note here is that I'm not intending to impose a contract on `mapStateToProps` and `mapDispatchToProps` but rather to make the existing expectations clear. In my IDE now I can hover over the `CalypsoifyIframe` import or instance and it will show me the list of props it expects.

In what ways would you be looking for it to be simpler? I think that like programming in general, there's lots of stylistic choices that we have available in typing systems and we can often type our code in different ways for a similar outcome.

mmtr replied:
> In what ways would you be looking for it to be simpler?

To me, it would be with a more verbose and easier code to read. So, having separate `StateFromProps` and `DispatchFromProps` could help with that.

However, I see how this can feel redundant and fragile.

I don't have a strong opinion about this, but in this case I'm bit more leaning towards readability rather than optimization.

jsnajdr replied:
> In what ways would you be looking for it to be simpler?

I don't know. The context is that I'm a TypeScript beginner and this single line introduces a lot of new concepts: `typeof` of a function, the `ReturnType` thing (BTW why is `ReturnType` used on one function and not on the other?), composing two types with `&` operator...

dmsnell replied:
> why is ReturnType used on one function and not on the other?

😄 

if you give it a moment's pause you might see why…`mapDispatchToProps` isn't a function. the fun part here is that if you try to use `ReturnType< mapDispatchToProps >` it will warn you of that.

> readability rather than optimization

while the point wasn't optimization the way I'd put it, the purpose was to express clearly what needs to be there.

I'm honestly curious @mmtr about your response. it's funny because in my head I don't care to know on that line what the types are - my concern is about communicating out about the React component in files where it's used. to that end I'm comfortable knowing that "the type of the connected props is the return type of the `mapStateToProps` function combined with the type of the `mapDispatchToProps` object.

what information in this line is confusing that you wish you knew? that is, what are you wanting from this line that it doesn't give you? are you wanting to see all the types inside the connected props? if you compare that to the way we have been doing it where those types simply aren't there, what is rougher about having that information available?

also, are you trying to infer everything from the text in the PR or did you checkout the branch and have the IDE show you the information? I don't think this stuff is nearly as useful if we don't take the IDEs into account as the purpose of these types is automation and communication.

> this single line introduces a lot of new concepts

I can explain what they mean, but I think your point is that there's new stuff. at least Microsoft has [good documentation](https://www.typescriptlang.org/docs/handbook/basic-types.html) for it as well as some intro tutorials. in the end I think we'll most commonly see a smaller subset of syntax in practice. we'll have to learn some new stuff _if we want to add typing_ but I want to stress in all this that we shouldn't feel obliged to introduce types if we aren't excited about it (on an individual developer/module/function/day basis).

 - `ReturnType< T >` means "use the type that the function `T` returns"
 - `typeof T` means "use the same type as T"
 - `T & U` is the type intersection. it means "use a type that combines the elements of the type `T` and also of the type `U`

TypeScript as a _structural_ type system stands in contrast to _nominative_ type systems meaning that it's permissive by nature - as long as a type meets the minimum requirements it will be fine. This isn't like Java where we have to also be of the same type name to work. This code shows some difference…

```ts
interface HasSpeed {
  speed: number;
}

const a: HasSpeed = { speed: 5.2 };
const b = { speed: 13.2, direction: 147 };

function distance( o: HasSpeed, ms: number ) {
  return o.speed * ms * 1000;
}

distance( a, 4.3 ); // obviously fine
distance( b, 3.1 ); // also fine, even though `b` isn't a `HasSpeed` it still meets the contract
```

One implication here is that we often see partial types or constructs that say "hey computer, _you_ figure out the types."

Again, to what purpose are we typing? I want to increase the safety of the code by having the computer automate bug discovery before shipping and I want to communicate to other devs what the expectations are for using my code. React components are of type `Component< P, S >` where we supply the `P` for prop type and `S` for internal state type. This line of code is saying, "TypeScript, join the inferred types of the connected functions and call that the connected props type"

---

Please let me know if my lengthy responses are introducing more confusion than helping.

dmsnell replied:
there are a few more generic "type wrappers" that come out of the box in TypeScript

 - `Readonly<T>` specifies that we can't change properties on the type `T`. This, for example, could be used to mark our state tree as read-only and prevent people from accidentally modifying `state` directly (I just saw someone do this maybe even last week in a PR where the write-path was obscured through a couple of variable assignments)

 - `Partial<T>` takes an object type and returns a new type where each property may be missing. it's a shorthand since we can do this by hand.

```ts
interface Vector { x: number; y: number, z: number }
interface PartialVector { x?: number, y?: number, z?: number }
interface PartialVector2 = Partial<Vector>;
```

like the example with `ReturnType<T>` we can see here that these mapping types allow us to keep our actual annotation footprint small and versatile. if we add a `t` element to the `Vector` interface then `PartialVector2` automatically gets it as well but `PartialVector` will be missing it. it's just like normal code except the values are types.

 - `Nullable<T>` just like `Partial<T>` except the properties have to be there, though they can be `null`

 - `Pick<T, K extends key T>`  this one allows us to constrain only a subset of type `T`

```ts
interface Site {
    siteId: number;
    title: string;
    url: string;
}

type SiteLike = Pick < Site, 'siteId' | 'title' >
```

![Screen Shot 2019-04-02 at 4 09 23 PM](https://user-images.githubusercontent.com/5431237/55442017-c1048b80-5561-11e9-9181-c9c4d208fac9.png)

 - `Parameters< T >` returns the type of function arguments for a function of type `T`

there are more of course. this might make a good learnup

mmtr replied:
> what information in this line is confusing that you wish you knew?

just like @jsnajdr said, I'm a TypeScript beginner and this line is introducing some concepts I'm not able to understand by just reading the code (unlike the rest of the PR).

Also, the fact that `mapStateToProps` and `mapDispatchToProps` are defined at the end of the file and not before this line is also contributing to my confusion.

> are you trying to infer everything from the text in the PR or did you checkout the branch and have the IDE show you the information? I don't think this stuff is nearly as useful if we don't take the IDEs into account as the purpose of these types is automation and communication.

yeah, that's true. I didn't checkout the branch and observed the changes on the IDE which definitely are useful there. But that's how I usually review code (I only checkout the branch to test locally, and then read the changes over the PR). 

> Please let me know if my lengthy responses are introducing more confusion than helping.

They are are useful for me. I love having these discussions 🙂. And let me remind what a wonderful job you did here!

jsnajdr replied:
Today I realized what was behind the gut feeling I had when seeing the `ConnectedProps`. It's the fact that the declaration doesn't contain any new information beyond what is already available in the source:
- the connected component accepts `Props` as input
- the `connect` function has a well-known type: it takes the `Props` passed from outside, adds the output of `mapStateToProps` and of `mapDispatchToProps`, and passes the resulting union down to the wrapped component.

Therefore, the fact that
```
ConnectedProps === Props & ReturnType<mapStateToProps> & ReturnType<mapDispatchToProps>
```
is an obvious universal truth about `connect`. Then why I need to repeat it in every file? With types for `connect` available, it should be inferred, shouldn't it?

jsnajdr replied:
> Please let me know if my lengthy responses are introducing more confusion than helping.

Your responses are very helpful, thank you!

dmsnell replied:
> It's the fact that the declaration doesn't contain any new information beyond what is already available in the source…an obvious universal truth

I've had to ponder this question and how to respond. There are a few ways to look at this. One way is that the type system is forcing us to do something slightly awkward or redundant. Another way to look at it is that we're matching the types with the way we code our components - composed pieces.

If we consider types as _sets_ this line should make sense as we walk through what's happening. But first, let's point out the _elephant in the room_: we can skip all this.

```ts
interface Props {
  siteId: number;
  postId: number;
}

export default connect(
	( state, { siteId, postId }: Props ) => ( { … } ),
	{ setPostTitle },
)( class extends React.Component {
	…
} )
```

With this formulation TypeScript should infer everything and we wouldn't need the redundancy. We could be writing our code like this but we separate out connected and unconnected components for things like testability. The implications of these separations bleed into the type system too as we have to or want to maintain the expectations on the unwrapped versions of the components.

The last commit where I added the `ProtectFormProps` can help illustrate this.

What is the _type_ of `CalypsoIframe`? _Because_ we use props from the `mapStateToProps` then we know that the type has to include those things beyonds what are defined in the public `Props` (or possibly we could call it `ExternalProps`) interface. The innermost component needs to constrain the input parameters to all things that it uses internally: the full set union of external props, connected props, and protected-form props.

I've created `ConnectedProps` as a convenience to group `connect()` elements but we could inline it just the same.

The `CalypsoIframe` component then is the set union of these types (for _reasons_ we use the intersection operator) or…
```ts
Component<ExternalProps & ConnectedProps & ProtectFormProps, InternalState>
```

Notice that we're _not_ adding any type annotations to `connect()`? Why? because we started on the inside and typed the inside and let TypeScript do the unwrapping. `protectForm` performs a kind of set difference on the types… (please excuse my made-up type syntax below)
```haskell
protectForm :: Component (P ∪ ProtectFormProps) -> Component P
```

That is, we're lifting out the subset of the type that is also `ProtectFormProps`. It leaves exposed `ExternalProps & ConnectedProps`. `connect()` does the same thing by removing the `ConnectedProps` though at the point it gets to `connect()` it's not using this convenience function but rather inferring them on its own.

We're not starting at `connect()` and then somehow injecting props into `CalypsoIframe`. Rather, we're starting with `CalypsoIframe` and the props it needs then using `connect()` to fulfill _some_ of them, leaving the rest to be supplied from the outside.

---

Let's backup and repeat in a different way:

 - `CalypsoIframe` actually needs things like `markChanged()` as provided by `protectForm()`. That means it _has_ to include that in its type.
 - Wrappers and higher-order components might _supply_ some props while _removing_ others. They do this through set operations on the types.

This is good news because when we do things like type `protectForm` the way I have here we get a clear type telling us what we should pass in on the imported object regardless of the internal implementation details, though we can still unwrap the internals for testing and non-standard composition.

We probably don't want `protectForm()` operating in our tests, but `this.props.markChanged()` is still called from inside the component. The inner type reminds us to provide mock functions for it in those cases.

![WrappedTypes](https://user-images.githubusercontent.com/5431237/55595654-c5fc4300-56f9-11e9-8805-3278698e898c.gif)

In this screencast we can see how it complains if we try and set `markChanged()` manually. While working this PR I discovered that if we _do_ pass in a value for `markChanged()` it will end up replacing the one built-in to the `protectForm()` wrapper. Is this a bug in the wrapper? I don't know, but now I am aware of that behavior and can enforce through the types not to do it.

If we pass in `siteId` to the connected component it will likewise complain, which is good, because that passed-in `siteId` will get overwritten by the one in `mapStateToProps`. This would be a latent bug in production otherwise and only expose itself when we have something happen like a mismatch in local and API data - maybe during the time an API call is resolving.

jsnajdr commented on client/gutenberg/editor/calypsoify-iframe.tsx:
```diff
@@ -38,28 +37,57 @@ import { getEditorPostId } from 'state/ui/editor/selectors';
+
+	constructor( ...args ) {
+		super( ...args );
 		MediaStore.on( 'change', this.updateImageBlocks );
```
This looks like a pre-existing bug. The `MediaStore.on( 'change' )` subscription should be created in `componentDidMount` (not in constructor) and removed in `componentWillUnmount` (not unsubscribed at all currently). Then the constructor can disappear.

dmsnell replied:
Great catch. For the missing warrant, I believe that we can end up calling the `constructor()` even in cases where the component never mounts, likewise we might run into situations where `constructor()` is called multiple times before mounting. Sound right?

I'll try and fix this bug outside of this PR but fold in the changes here. That's a behavioral change I'd like to isolate.

jsnajdr replied:
> For the missing warrant, I believe that we can end up calling the constructor() even in cases where the component never mounts, likewise we might run into situations where constructor() is called multiple times before mounting.

Yes, with React Fiber, error boundaries and concurrent mode, it's increasingly common to perform some work (constructing a component), then throw it away for some reason (error thrown from render and caught by error boundary, high-priority concurrent update preempting an already running one, suspending a render) and then restart it again.

`constructor` doesn't have an associated destructor, so acquiring resources in it can lead to leaks.

jsnajdr commented on client/gutenberg/editor/calypsoify-iframe.tsx:
```diff
@@ -152,38 +185,46 @@ class CalypsoifyIframe extends Component {
+		post_title: string;
+		post_excerpt: string;
+		post_content: string;
+	} ) => {
```
Creating a `PostRevision` type would let us avoid this unwieldy destructuring.

dmsnell replied:
I thought about it but felt like creating a `PostRevision` would both be too restrictive and something I don't know if I need. Were we to have a full `PostRevision` model, or if I knew it, I would have reused it here. that being said, we only need these three props so I left the restructuring in.

jsnajdr commented on client/gutenberg/editor/calypsoify-iframe.tsx:
```diff
@@ -210,29 +251,34 @@ class CalypsoifyIframe extends Component {
 	};
 
-	updateImageBlocks = action => {
+	updateImageBlocks = ( action: { data: { mime_type: string; URL: string }; type: string } ) => {
```
Having a separate typedef for this action would be more readable than the inline type.

dmsnell replied:
I'm not sure what to do in these cases. I guess we can create types or interfaces for them, but I like having these things inline so I don't have to jump around to find them. much of my trouble here has to do with the fact that I don't know what these should be so I'm going for minimum effort on typing. I'll think about ways to do this otherwise

jsnajdr replied:
The Happy Tools folks use Flow to typecheck their Redux code. They have typedefs for actions that allow type inference by the `type` value:
```js
switch ( action.type ) {
  case 'UPDATE_IMAGE_BLOCKS':
    // at this place your editor knows what fields the `action` has
}
```
See p9lV3a-r7-p2. That could be a way to typecheck Redux with TypeScript, too.

dmsnell replied:
thanks @jsnajdr. it's true we have several options here to type on. I guess you are saying that this is a Redux action - I didn't realize that at the time.

typing actions is a broader question and one which will undoubtedly push into more advanced type annotations. it's entirely possible though.

I'll look into that action and see what I can come up with. in some cases I think I'd still fall back to providing a minimal type until we have more of the core modules typed. that is, getting our actions typed might be a huge project, but I can still let the callee know what I need and expect here with these few parameters.

we can do a funky mid-term solution here as well though I wouldn't personally introduce this.

```js
( action: { … }, type: 'UPDATE_IMAGE_BLOCKS' as const ) => { … }
```

what it brings us is a warning at the call site if we try and send any action of a different type. now, a benefit of using the `@babel/typescript` preset is that it won't refuse to compile if we have this type error there. `tsc` _would_ refuse until we could provide ample evidence that only that string literal can be passed in as `type`

jsnajdr commented on client/gutenberg/editor/calypsoify-iframe.tsx:
```diff
@@ -252,14 +298,20 @@ class CalypsoifyIframe extends Component {
+	}: {
+		postUrl: string;
+		previewPort: MessagePort;
+	} ) => {
```
The `postUrl` and `previewPort` are not really meant as fields in a record type, but rather named/optional parameters of the `openPreviewModal` method. Does TypeScript have some nice way how to express that? I.e., without writing `postUrl` twice?

dmsnell replied:
to my knowledge it doesn't make an opinion on this. its our choice to wrap them in an object where it gets weird. the syntax for annotating types is the same syntax as for destructuring renames

```ts
( { postUrl: string } ) => ()
```

In this example are we trying to rename `postUrl` to `string` or declare a string type? the solution is to type the object itself separately as is done above.

---

> not really meant as fields in a record type

The types here aren't trying to convey that they are such. TypeScript is _structural_ and not _nominative_ so all we're trying to convey is that "this function needs an argument that's an object and it needs these things in it."

The way to remove the verbosity is to eliminate the _arguments as an option bag_ approach (which is something I'm all for), though in all fairness, if I could type something twice and eliminate an unexpected bug in my software while shaving off time for other developers who are wanting to know what my function needs, I'll take that trade.

In the meantime, I'll remove the record syntax; these are required parameters anyway. Without types the record syntax can be a nice way of adding some sense of types, though at the cost of runtime allocations.

dmsnell replied:
Replaced in f84adf0486

<img width="684" alt="Screen Shot 2019-04-01 at 10 55 12 AM" src="https://user-images.githubusercontent.com/5431237/55348738-c03dfd80-546c-11e9-875f-2915ec926bc1.png">

No information in my IDE was lost by changing `( { postUrl, previewPort } )` into `( postUrl, previewPort )`

jsnajdr commented on tsconfig.json:
```diff
@@ -0,0 +1,26 @@
+		"moduleResolution": "node",
+		// Process & infer types from .js files.
+		"allowJs": true,
+		"jsx": "react",
```
If we start compiling the `.tsx` files with the real TypeScript compiler, which I see as the next step, how do we want our pipeline to look like?
1. `tsx` directly to `js` that will be bundled with webpack, with no intermediate Babel step. Then we need the `jsx: react` option and `target: esnext` will be different for evergreen/ie11 builds.
2. `tsx` to `esnext` (optionally with `jsx`), will be further transformed by Babel before bundling.

As long as we have custom Babel transforms (nonstandard JSX for `wp.element`, inline Redux action imports, `AsyncLoad` transform, ...), we need the intermediate step.

dmsnell replied:
> the next step, how do we want our pipeline to look like?

This used to be the source of my biggest questions @jsnajdr - that's one reason why I never felt confident in pushing #15390. Compiling with `tsc` provides a number of significant perks (browser compatibility, speed, a native way of robustly handling packages in a monorepo) but leaves a few unanswered questions.

For one I think there are two dominant TypeScript loaders and I'm not clear on the tradeoffs. Maybe we just pick one of them and we'd be fine.

Second, those babel transforms are tricky. Honestly though I think we could consider removing the action type inlining. If I remember correctly that shaved 10 KB off of the bundle sizes, but I would probably be willing to give that up if it let us use TypeScript as a real compiler. Who knows, it might be possible that we could gain that 10 KB back without too much hassle through better code generation.

Concerning `wp.element` I'm wondering how big of a problem that is. Could we find a simple way to exclude those files if indeed it became a problem? I don't know the answer here.

---

If we could eliminate our custom transforms, which seems like a reasonable possibility (doesn't `AsyncLoad` provide functionality which became standard with `import()`?) then I think a great end-goal is eliminating Babel. I don't see it being practical at this moment but maybe we could envision a time when it would be. (`tsc` has no problem compiling in `.js` files)

sirreal replied:
> Concerning wp.element I'm wondering how big of a problem that is. Could we find a simple way to exclude those files if indeed it became a problem? I don't know the answer here.

As far as I know, we don't have any builds that require a mix of `React.createElement` and `wp.element.createElement`, so we might be able to just make sure that whatever build is running is correctly configured.

jsnajdr replied:
> doesn't `AsyncLoad` provide functionality which became standard with `import()`?)

`AsyncLoad` is certainly ripe for a major update. When it was created, the ESM dynamic `import()` didn't exist, we used CommonJS modules, and `React.lazy` wasn't a built-in React feature. It's entirely possible that after the update, the custom transform won't be needed any more.

jsnajdr replied:
> Honestly though I think we could consider removing the action type inlining.

Some time ago I had an idea that it could be a webpack plugin instead. The plugin would detect that a module export is a string/numeric constant and would inline it at the import site. The module concatenation and the `sideEffect: false` plugins are conceptually similar: they also analyze the import/export graph and eliminate indirection.

That would make it a linker-time optimization instead of a compiler-time one.

jsnajdr replied:
> For one I think there are two dominant TypeScript loaders and I'm not clear on the tradeoffs.

Webpack and TypeScript docs describe `ts-loader` as the first choice. The other one is `awesome-typescript-loader`, is that right? They mention good Babel integration as their main advantage over `ts-loader`. And that sounds like something that we might like 🙂

jsnajdr replied:
> As far as I know, we don't have any builds that require a mix of `React.createElement` and `wp.element.createElement`

This can change quickly as soon as we start importing anything from `@wordpress/components`. Better be prepared 🙂 Not a blocker right now, just let's not create something that will be a blocker in future.

blowery replied:
> This can change quickly as soon as we start importing anything from @wordpress/components

If we import from there, the code will probably already be transpiled down to wp.element calls.

jsnajdr reviewed (commented):
Thanks Dennis for being persistent with types! :wink: I think it's no longer reasonable to block the introduction of a type checker.

My biggest fear though is that contributors will think that they have a duty to introduce types absolutely everywhere. That will lead to a Java-style verbosity with very little benefit in return.

Even this very first PR includes a few examples that make me 🙁: one untyped line becomes 10 typed lines and I'm not sure if that made the program better or less buggy.

Regarding the Babel plugin/preset: what are the pros and cons of the `@babel/typescript` preset versus the full official compiler? I'm a complete beginner in that area.

jsnajdr commented on babel.config.js:
```diff
@@ -43,7 +44,7 @@ const config = {
 	env: {
 		test: {
-			presets: [ [ '@babel/env', { targets: { node: 'current' } } ] ],
+			presets: [ [ '@babel/env', { targets: { node: 'current' } } ], '@babel/preset-typescript' ],
```
This one is not needed. In the `env` config, we only specify additional presets that are not present in the base config.

jsnajdr commented:
> I have very little confidence that I properly setup the babel config. I'm confused by the indirection in our babel configs now and don't understand where I need to put the transform.

Just adding a new `@babel/typescript` preset to the base config should be enough. It's just another transform, and surprisingly, one of the more simple ones. It just removes the type nodes from the AST and expands the enums into their values.

dmsnell commented:
> one untyped line becomes 10 typed lines

not sure if you are referring to the function arguments where restructuring required duplicating the param names or not. that's about as bad as I expect it to get. in one case at least you tossed out the idea of creating a separate interface and I think that's valid, but in the spirit of only typing what's required to communicate and provide safety I left it inline (didn't know what the type should be)

I expect lots of experimentation on how we type and some extra verbosity, though I don't expect Java-like verbosity. even in the calypso iframe file most of the 400 lines were untouched.

> contributors will think that they have a duty to introduce types absolutely everywhere

that's a valid fear. I'm in a better spot now with my priorities to help write and document more and you raise a good point that we need to educate each other on how to type things.

example: with `EditorActions` I _chose_ to create an `enum` type _in order to_ communicate what the possible string values are that we can handle and to make it more difficult to overlook one. that is a choice that could have been left out. in `mapDispatchToProps` I left out any typing and instead relied on TypeScript to infer what it can.

the act of reworking this one file even taught me a few things that I didn't anticipate running into, though those things were exciting to me. for instance, we already try to add types in a few places like with `propTypes`, but in this file we were missing `props` that we were using, so the verbosity was helping me complete a job that probably should have been done before.

anyway I don't mean to ramble. I hear your point and I think a little extra effort watching changes in `.ts` and `.tsx` files paired with some learn-ups might help alleviate that.

---

> what are the pros and cons of the @babel/typescript preset versus the full official compiler?

sorry I missed this comment earlier @jsnajdr when I reviewed it. I tried to answer some in [another comment](https://github.com/Automattic/wp-calypso/pull/31902#discussion_r270979189) but the summary from my impression is that by using the babel preset we can avoid some complicated design-related question around the build pipeline, something that plagued me in the past when trying to bring this in.

the biggest drawbacks I think are these:
 - we're still relying on Babel, which I expect to be less robust and much slower
 - we can't take advantage of ecosystem perks like [TypeScript's Project References](https://www.typescriptlang.org/docs/handbook/project-references.html) which I don't understand but think are designed to solve issues in big monorepos (what `lerna` does for us except in an elegant and efficient way)
 - there are a few nice TypeScript features not present: namespaces and `const enum`s. the `const enum` is something I would have liked to use in this module because it inline-replaces strings. in fact, a `const enum` is the same thing as what our action-types are _with our custom babel plugin_. a standard `enum` still creates the object with the named choices and performs a lookup at runtime. so `EditorActions.TrashPost` is basically what you get instead of a replacement with `trashPost`

what we gain?
 - everything else stays the same
 - we can still let type errors through since Babel is compiling. this means that it's easier to add types to part of an existing module without fixing the whole thing

basically this is the smoothest way we can introduce this without forcing a big change.

dmsnell committed:
Introduce TypeScript and type Calypso Block Editor iFrame bridge

See prior work in #13117, #15390, and #15397

This PR introduces TypeScript into our Babel presets so that we can write in TypeScript and take advantage of the safety and development experience improvements that it offers.

I have updated the _iframe bridge_ for the block editor as a first transformation to not only demonstrate what an update looks like but also because the integrity of that module is relatively important and the introduction of types brings some immediate tangible benefits.

dmsnell commented on client/gutenberg/editor/calypsoify-iframe.tsx:
```diff
@@ -38,30 +37,55 @@ import { getEditorPostId } from 'state/ui/editor/selectors';
-		this.mediaSelectPort = null;
-		this.revisionsPort = null;
-	}
+	iframeRef: React.RefObject< HTMLIFrameElement > = React.createRef();
```
this line isn't entirely necessary as written because I could have simply put `iframeRef = React.createRef()` but by typing the `HTMLIFrameElement` it not only specifies what it will refer to but opens up the further inference and autocomplete on those properties unique to the iframe, such as `contentWindow`, which may not be there in general `HTMLElement`s

jsnajdr commented:
Here's a was how to incorporate the official TypeScript compiler in what I think is seamless and elegant way:
1. Transpile `.tsx` files with `ts-loader` and `target: esnext` and `react: preserve` options. That will produce ES2018 code with ESM imports and exports and JSX markup. Basically the same type of code that we write by hand in `.jsx` files.
2. Transpile the `ts-loader` output with Babel, like any other `.jsx` file. That will transpile JSX with optional `wp.element` support, transpile `<AsyncLoad>` instances, and optionally transpile to ES5 if not inside evergreen build.

The advantage is that JSX and ES5 will be processed only at one place (Babel) instead of two. Easier to maintain and debug.

Disadvantage might be slower build -- compared to the Babel-less, TS-only build.

Maybe we will eliminate Babel in the future, but I don't see that coming anytime soon. And I personally have a positive attitude towards Babel. In the recent months, we simplified and cleaned up the config a lot, and it will get even simpler over time. I don't think that "one of a million flavors of JS + proposals + browsers + JSX" is a fitting description of our current Babel reality.

obenland commented:
> It's difficult for me to enumerate many drawbacks. The benefits here are significant for [...] developer experience.

I've never looked at Typescript so far, so maybe I can help here 🙂 

By adding Typescript support we'll add another layer (and multiplier) of complexity. I'm not sure how many community contributions we receive for Calypso, or how much work we put in to encourage them, but with this added layer of complexity I'd anticipate **community contributions to suffer**.

We'll require Automatticians to learn up on it, adding to the **cost of educating existing contributors**. I'd also expect it to have an effect on **contributor onboarding**.

As beneficial stronger typing is to developer experience, you could make the argument that having to manage yet another code transition is not (I heard recently that we still use Flux in parts of Calypso).

I can't adequately judge the merit or need of stronger typing in Calypso, but as a novice contributor (to Calypso) I know it would make me more hesitant to dive into `.tsx` files and less confident about my work.

If we are to go forward with it, excessive documentation would be a must in my opinion. I'm not sure how many Automatticians would currently be able to describe what's happening here:
```typescript
type ConnectedProps = ReturnType< typeof mapStateToProps > & typeof mapDispatchToProps;
```

dmsnell commented:
> seamless and elegant way

@jsnajdr having tried both of the loaders _and_ the preset _and_ separate compilation altogether this is going to stay my recommendation. I'd actually be happier to rely more on `tsc` for compilation but I haven't found any good ways in my experiments. I'll try and create a branch from this PR and replace the preset with the loader - maybe I'm jaded by the state of the loaders in the past.

in the end though, if `tsc` forces us to fix all of our errors before compiling that feels like a blocker. I expect type adoptions to be very gradual as we all learn and my expectation is that most of us will encounter types in a read-only fashion, that is, through autocomplete and inline warnings in the IDE.

> By adding Typescript support we'll add another layer (and multiplier) of complexity.

I hear you @obenland and I sympathize. I don't have much to say but to ask to be open-minded about this and extend a little trust. I believe that introducing types will have a net positive effect even if there are ways that it raises the barrier. why do I think this? because I think that it has the potential to dramatically lower the barrier in other ways.

for example, I can share one of my own frequent frustrations: "what does this code expect?" and I'm left with no help. someone recently shared this on the WordPress Slack team in the context of Gutenberg.

> Is there a list of available UI icons for Gutenberg anywhere? I notice several of the icons don’t match the Dashicons. Specifically I’m looking for a list of available icons to use with the `icon-button` and similar component (_sick_) [in #core-editor](https://wordpress.slack.com/archives/C02QB2JS7/p1553880176066200)

particularly with things like configurations that carry optional values and categories/icons/types/etc… I know that being able to hit <kbd>Ctrl</kbd>+<kbd>Space</kbd> (or <kbd>⌘</kbd> on macOS) would immediately help me orient myself and take away the uncertainty which leaves me scanning other source code to try and find examples sufficient enough to answer my question.

maybe I'm alone, but I frequently find myself having to jump around and read multiple levels of source code to answer a basic question: which arguments and of what type and of what possible values does this function expect and return?

so we have a mixture of reasons motivating the addition of types: on one hand we can eliminate uncertainty in our codebase and hopefully people find that enough of a help that it covers the need to learn some new syntax; on the other hand we can uncover errors that our customers experience and in my opinion that's worth something too even if it makes some things slightly harder on the developer if it makes for a less frustrating experience for the customer.

stepping back though I know Calypso already has an enormously high barrier to entry. I think the lack of our ability to allow the computer to assist with types and bug-finding is part of that. of note, we've introduced many different languages over the years and it has gone okay, even if not idea: EventEmitters, Flow, React, Redux, Promises, static import/export statements, etc…

---

It's fine to use plain JavaScript inside a `.ts` file. Considering the concern about learning some new material, do we think there's a way to let people know that it's fine to just continue programming in the JavaScript they are used to?

Definitely blog posts, maybe an entry in the `README` and in the coding docs. I can try and add some content here in this PR

sirreal commented:
> Existing JavaScript continues to work, though we can run tsc to type-check existing code.

I'm very interested in this point and had hoped to explore the possibilities myself. If I understand correctly, this means that by typing some areas of the codebase we can get additional safety beyond just the `.ts(x)` files? Have you considered areas where this would be especially beneficial?

- Critical systems (checkout)
- Widely reused code (site selection, data-layer)

dmsnell commented:
> I'm very interested in this point

It's a huge step-up for gradual adoption in my opinion @sirreal. If you checkout this branch and then `npm install` you should be able to start seeing it right away. `PhpStorm` worked without configuration as did `VSCode`. Many of the IDEs check for the presence of the `tsconfig.json` file and that's all they need.

> Have you considered areas where this would be especially beneficial?

The block editor iframe bridge in this file is one of those places I consider more critical to get right. We have messaging happening which requires lots of accounting to get right. Some of that is mitigated by custom `MessageChannel` ports, but it's important we uncover things like "the port may not be available right here" in anticipation of broken flows for our customers.

We have **a bunch of utilities** that are widely used that would gain from this too. Without any further work or knowledge on the part of most contributors those utilities could become more clear by typing them. Heck, in this PR I'm automatically benefiting through the inference of the types of the action creators.

I'd love to eventually **type our state too**. Much of our fragility comes in through assumptions about properties that exist in state which are wrong (or more accurately, _sometimes wrong_). Additionally there are a number of common selectors which could actually be removed long-term. Though some selectors do things like cache recent values, plenty of them are there solely for pseudo-runtime-based type contracts. If we could cleanly connect the shape of state with the code consuming it then we could remove those, shrink the build, speed up the runtime, and still have higher confidence when making changes to that state code.

Finally if we really wanted to go whole hog we could start pulling back our JSON-schema validators for API calls. Much of our schema is there again to provide a faux runtime-based type system. Having **_types_ on the API data** would bleed far into the app. When was the last time you accessed a data property on an API call response that ended up not being there or ended up being wrong? There's still a place for the parsers in the data layer, or even more of a place because now those parsers can return the constrained types and that can bleed deep into the app. I really wanted types on the comments data we used in the comments management section and it just wasn't happening - everything is a big bag of properties and good luck knowing what they are and what properties are appropriate.

I'm sure there are more. Lots of React code should stay the same. I expect lots of components to benefit from the way that the types get inferred and passed in. Right now, for example, we have to manually constrain our action-creator calls; with typing in there we not only know what to pass, but when we make function API changes we can follow the type errors to make sure we don't break any existing consumer sites.

jsnajdr commented on client/gutenberg/editor/calypsoify-iframe.tsx:
```diff
@@ -353,7 +390,10 @@ class CalypsoifyIframe extends Component {
-const mapStateToProps = ( state, { postId, postType, duplicatePostId } ) => {
+const mapStateToProps = (
+	state,
+	{ postId, postType, duplicatePostId }: { postId: number; postType: string; duplicatePostId: any }
```
The type of this argument is the already-defined `Props`. I.e., the props passed from outside to the connected component.

dmsnell replied:
oooh. good catch. this is interesting. I'll update to reflect that.

`connect()` does some [fun advanced typing](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/react-redux/index.d.ts#L113) to produce the right external type, but `mapStateToProps` isn't covered by that.

jsnajdr commented on client/gutenberg/editor/calypsoify-iframe.tsx:
```diff
@@ -202,17 +238,13 @@ class CalypsoifyIframe extends Component {
-	closeMediaModal = media => {
-		if ( ! this.state.classicBlockEditorId && media ) {
+	closeMediaModal = ( media: { items: Array< Parameters< typeof mediaCalypsoToGutenberg > > } ) => {
+		if ( ! this.state.classicBlockEditorId && media && this.mediaSelectPort ) {
```
The `this.mediaSelectPort` check is interesting. Did the typechecker force you to add it? Because it shouldn't be needed, for the following reasons:
- `closeMediaModal` can be called only when the Media Modal is rendered, i.e., `this.state.isMediaModalVisible` is true
- `this.state.isMediaModalVisible` and `this.mediaSelectPort` are set and unset in sync, when opening and closing the modal.

The fact that one of the properties is set on instance and the other on state, and that the state update can be scheduled for later rather than done immediately, looks like a weak point of the program.

Also, the `isMediaModalVisible` property is redundant, as it can be derived from `this.mediaSelectPort`.

Is it possible to reorg the program in such a way that the typechecker "knows" the above facts? That would be better use of the type system than just adding checks that are not really needed. It's similar to patching ESLint a11y warnings in such a way that the linter shuts up, but the actual a11y remains bad.

dmsnell replied:
> Did the typechecker force you to add it?

yes it did. you make a valid claim about it and it's a case where I like to argue with the type checker. to its credit though, we occasionally get code refactors that change an assumption that currently holds. in these cases it provides a guard against unintended side-effects of other changes.

> Is it possible to reorg the program in such a way that the typechecker "knows" the above facts?

This is great! It's an advanced type of type-driven design that seeks to remove as much information as is possible; in this case it removes the opportunity for `this.mediaSelectPort` and `isMediaModalVisible` to get out of sync.

I may try and get this change in outside of this PR as it seems like a good one. I'll double-check though to be sure.

dmsnell replied:
good idea @jsnajdr but I don't think we can simplify yet. the classic block opens the media modal but doesn't pass a `MessagePort` so in the case that we open the media model from the classic block `this.mediaSelectPort` will actually be `null`

kwight commented:
I absolutely <3 the idea of Calypso adopting TypeScript – I'm certain that two years down the road, 100% of us would feel the returns have been 10x. In the meantime, it'll be frustrating (and feel time-consuming) for many of us, for all the reasons we're all already aware of 🙃 . That said, a few things off the top might help adoption in the long run.

By the way, I love how this is implemented; we're not forcing it on anyone in any way. If a quick fix needs to be done, a dev can shove good ol' JS into a `.ts` file and merge away (if I understand correctly). I think this is key to letting devs pick it up on their own terms, rather than it being dictated in stressful ways.

I'd suggest a basic README be added to this PR. It could have a few links to getting set up and getting started, a small FAQ (basic ideas, like, "mostly you just have to say what type of data parameters are, and what the functions returns; the expressions within a block mostly stay the same!"), and something explaining what a strongly-typed file of code "looks" like. For example, when unfamiliar with strongly-typed languages, a file looks incredibly confusing and verbose; it  looks like "normal" code, with all sorts of extra noise sprayed everywhere. I don't know, maybe there are existing TypeScript tutorials and videos that explain this already; even just pointing to those would be a big help.

I'd be happy to help (I'm not familiar with TypeScript at all, but have been getting more comfortable with strongly-typed languages by playing with Dart lately). Either I can give feedback on a README, or do one up myself if time isn't of the essence.

dmsnell commented on client/blocks/comments/form.jsx:
```diff
@@ -19,7 +19,7 @@ import { getCurrentUser } from 'state/current-user/selectors';
 import { recordAction, recordGaEvent, recordTrackForPost } from 'reader/stats';
 import { isCommentableDiscoverPost } from 'blocks/comments/helper';
-import { ProtectFormGuard } from 'lib/protect-form';
+import { ProtectFormGuard } from 'lib/protect-form/index.tsx';
```
I wish I understood why this change was necessary and how to prevent it.

jsnajdr replied:
It's the server build that crashes here:
```
ERROR in ./client/blocks/comments/form.jsx
Module not found: Error: Can't resolve 'lib/protect-form' in '/Users/jsnajdr/src/wp-calypso/client/blocks/comments'
 @ ./client/blocks/comments/form.jsx 36:19-46
 @ ./client/blocks/comments/form-root.jsx
 @ ./client/blocks/conversations/list.jsx
 @ ./client/blocks/reader-post-card/conversation-post.jsx
 @ ./client/blocks/reader-post-card/index.jsx
 @ ./client/components/seo/reader-preview/index.js
 @ ./client/components/seo-preview-pane/index.jsx
 @ ./client/components/web-preview/content.jsx
 @ ./client/components/web-preview/component.jsx
 @ ./client/components/web-preview/index.jsx
 @ ./client/my-sites/themes/theme-preview.jsx
 @ ./client/my-sites/theme/main.jsx
 @ ./client/my-sites/theme/controller.jsx
 @ ./client/my-sites/theme/index.node.js
 @ ./client/sections.js
 @ ./server/pages/index.js
 @ ./server/boot/index.js
 @ ./index.js
```
You need to add the TypeScript extensions also to the `webpack.config.node.js`.

dmsnell replied:
ahhhhh!!!! thanks!!!

sgomes reviewed (commented):
Putting on my developer relations hat from a previous life, I think the approach described is good. I could see a lot of pushback from the community if a new way of doing things were mandated, particularly when it would involve adopting what is ostensibly a different programming language. 

Therefore, I'm glad this is being framed as a learning experience and a way of adding extra safety to more critical areas of the codebase. Developers motivated to work on those critical pieces would likely be willing to learn as much as they need to complete their work, but if not I'd still like to see their code being accepted, with some gentle nudging to consider type safety, or perhaps a future PR by someone else adding it back.

With my developer hat on, I am neutral to the idea of type-safety. I believe type-safe and non-type-safe development are both perfectly valid, but they require different testing approaches, the latter relying on more extensive unit and integration testing, and ideally the use of tools like fuzzers.

I do think that guidelines, documentation and perhaps extra linting are needed as part of this rollout, though, so that those of us who aren't familiar with TypeScript but want to develop in it can get started on the right foot. I'd also suggest creating a reviewer group composed of TypeScript-savvy folks, so that we could add that group to the list of reviewers to get a language/style review, beyond the usual review process. That would help ensure that our TypeScript code style evolves in a consistent direction, rather than becoming a patchwork of different practices, whether equally valid or not.

Thanks for championing this effort, @dmsnell ! 🎉

sirreal commented:
I've pushed a fix for the failing tests where `ts` and `tsx` files were not being transformed by `babel-jest`.

sirreal reviewed (dismissed):
I've pushed a few changes to prepare our tooling for tsx.

Lint TypeScript files by our precommit hook and on CircleCI. There are a lot of errors and warning introduced: https://circleci.com/gh/Automattic/wp-calypso/250648

`plugin:@typescript-eslint/recommended` introduces a different set of rules, we should make sure they make sense in Calypso before enabling them. There's a risk of a new set of rules being quite disruptive for developers. It's a good idea to take a look at new results from linting. `npm run lint:js` is extremely slow, so I've looked at some samples like `npx eslint --ext .js --ext .jsx --ext .ts --ext .tsx --cache client/components`. I've updated eslint config to disable or configure the most problematic rules from the new preset to better align with our existing config.

**While updating some package scripts to handle TypeScript, I realized that our translation tooling is not prepared for TypeScript.** See https://circleci.com/gh/Automattic/wp-calypso/250766

This PR doesn't introduce any translated strings in TypeScript, but we should be able handle them before enabling TypeScript in Calypso. This will likely require ~3 changes to parts of our tooling:

- Update [xgettext-js](https://github.com/Automattic/xgettext-js) to handle TypeScript and publish new version
- Update [i18n-calypso](https://github.com/Automattic/wp-calypso/tree/master/packages/i18n-calypso/cli) to use the TypeScript enabled `xgettext-js`
- Land https://github.com/Automattic/gp-localci-client/pull/16 and Revert 3675056

jsnajdr commented:
> Update `xgettext-js` to handle TypeScript and publish new version

It should be sufficient to update the `i18n-calypso` tool by adding `plugins: [ 'typescript' ]` to the [`parserOptions` config](https://github.com/Automattic/wp-calypso/blob/d90bc1b5ffe7164810161b17b1331c5d45ffa565/packages/i18n-calypso/cli/i18n.js#L36-L46). Then the `@babel/parser` will understand TypeScript syntax, without any further changes.

dmsnell commented on packages/eslint-config-wpcalypso/index.js:
```diff
@@ -4,7 +4,8 @@ module.exports = {
 	},
-	extends: 'eslint:recommended',
+	extends: 'plugin:@typescript-eslint/recommended',
+	parser: '@typescript-eslint/parser',
```
not something I have much of an opinion on. as long as Calypso and all its derivatives build I'm cool with whatever I imagine we would do.

these configs are areas I'm less familiar with so you can assume less intent with my code than in other places, such as in adding types to files

dmsnell commented:
> plugin:@typescript-eslint/recommended introduces a different set of rules, we should make sure they make sense in Calypso before enabling them. 

that's a salient point @sirreal. I backed out the overrides you added and removed the whole recommended plugin. feel free to add your rules back in or tell me to do so; I was thinking that maybe at this point we let things form a bit before we start locking down rules.

sirreal commented on .circleci/config.yml:
```diff
@@ -238,7 +238,7 @@ jobs:
           when: always
           command: |
-            git clone --single-branch --depth=1 https://github.com/Automattic/gp-localci-client.git
+            git clone --single-branch --depth=1 --branch 'update/include-typescript' https://github.com/Automattic/gp-localci-client.git
```
I opened #32352 to isolate the i18n changes and unblock https://github.com/Automattic/gp-localci-client/pull/16

kwight commented on docs/coding-guidelines/typescript-dev-flow.dot:
```diff
@@ -0,0 +1,34 @@
+    existing_file[label="opening an existing file"];
+    am_comfortable[label="am I comfortable\nworking with strong\nstatic types?"];
+    no_types[label="don't sweat it.\nwrite normal JavaScript in the\nfile as much as you can."];
+    learning[label="add as much or little TypeScript as you want\n\nthe types tribe will review your work\nand provide feedback to help you learn"];
```
Can you rename the "types tribe" references to maybe "type review team", something that better aligns with the team we would tag for help?

kwight replied:
Also picky: can you capitalize the first letter of sentences?

dmsnell replied:
updated and rebased so the actual commits are somewhat hidden

kwight commented on docs/coding-guidelines/typescript.md:
```diff
@@ -0,0 +1,300 @@
+TypeScript Coding Guidelines
```
This doc is absolutely fantastic, @dmsnell – it's everything I suggested in the vaguest way possible as being really helpful. The `Avoid / Prefer` model is super helpful. <3

wangkailang commented:
Why not publish support for typescript to _NPM, any legacy work ?

sirreal commented:
> Why not publish support for typescript to NPM, any legacy work ?

@wangkailang Your question is unclear for me. Can you elaborate?

wangkailang commented:
> > Why not publish support for typescript to NPM, any legacy work ?
> 
> @wangkailang Your question is unclear for me. Can you elaborate?

Oh, [npm published 4.0.0-alpha.1](https://www.npmjs.com/package/i18n-calypso/v/4.0.0-alpha.1) has support, but [CHANGELOG](https://github.com/Automattic/wp-calypso/blob/master/packages/i18n-calypso/CHANGELOG.md) not record changes, maybe need update.

sirreal commented:
> Oh, [npm published 4.0.0-alpha.1](https://www.npmjs.com/package/i18n-calypso/v/4.0.0-alpha.1) has support, but [CHANGELOG](https://github.com/Automattic/wp-calypso/blob/master/packages/i18n-calypso/CHANGELOG.md) not record changes, maybe need update.

You want to use `i18n-calypso` to parse TypeScript? You're right, support was recently added but I don't believe we've released that yet. I'll try to get it released next week 👍

sirreal commented:
@wangkailang Keep an eye on https://github.com/Automattic/wp-calypso/pull/35600. I hope to do another i18n-calypso pre-release `i18n-calypso@4.0.0-alpha.2` that will include TypeScript support. Let's move this conversation there 🙂
